dashboard/test/models/level_test.rb

```diff
@@ -464,6 +464,34 @@ def create_maze
     assert_equal 'PlantASeed', level.name
   end
 
+  test 'cannot create two blockly levels with same key' do
+    game = create :game, name: 'Artist'
+    level1 = create :artist, name: 'blockly', level_num: '5_5_5', user_id: nil, game: game
+    assert_equal 'blockly:Artist:5_5_5', level1.key
+    e = assert_raises ActiveRecord::RecordInvalid do
+      create :artist, name: 'blockly', level_num: '5_5_5', user_id: nil, game: game
+    end
+    assert_includes e.message, 'Validation failed: Level num has already been taken'
+  end
+
+  test 'cannot create multi and match levels with same key' do
+    multi = create :multi, name: 'multi match', level_num: nil, user_id: nil
+    assert_equal 'multi match', multi.key
+    e = assert_raises ActiveRecord::RecordInvalid do
+      create :match, name: 'multi match', level_num: nil, user_id: nil
+    end
+    assert_includes e.message, 'Validation failed: Name has already been taken'
+  end
+
+  test 'cannot create custom and dsl levels with same key' do
+    level1 = create :artist, name: 'artist multi', level_num: 'custom'
+    assert_equal 'artist multi', level1.key
+    e = assert_raises ActiveRecord::RecordInvalid do
+      create :multi, name: 'artist multi', level_num: nil, user_id: nil
+    end
+    assert_includes e.message, 'Validation failed: Name has already been taken'
+  end
```

Comment:
I was not able to get this test case to fail without the changes in this PR, due to the existing line: https://github.com/code-dot-org/code-dot-org/blob/a393f961add84e034fcd59b539ad769d24431946/dashboard/app/models/levels/level.rb#L56 I was not able to pinpoint why this is, so I'm relying on inspection to make sure that this test is actually testing the right thing.

Reply:
It looks like the Blockly level factory assigns a user id https://github.com/code-dot-org/code-dot-org/blob/e6601d6b7c98cf08f05b167153f88480df366917/dashboard/test/factories/factories.rb#L496

Reply:
true, but that doesn't totally explain it for me because I'm specifying `user_id: nil` on the multi level.